### PR TITLE
💥 Add Presign and Deployment Script

### PR DESCRIPTION
**Pull request by pcaversaccio**
### 🕓 Changelog

This PR adds an `ethers`/`hardhat`-based pre-signing and deployment script for the contract creation transaction of `CreateX`.

It's as simple - using a personal test account - as the following (another way of course to be documented later is by using `cast`):

image

image

Example deployment: [`0x9a30661B895f94d4A75Bd438E98b5Ae219620d50`](https://base-sepolia.blockscout.com/address/0x9a30661B895f94d4A75Bd438E98b5Ae219620d50?tab=contract)

> PS: This obviously only works for RPCs / networks that support pre-[EIP-155](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md) transactions.

#### Other Changes
- Bump `prettier` from `3.0.3` to `3.1.0` (https://github.com/pcaversaccio/createx/pull/31/commits/c309ea08f5f7e5b6f92fd1e5f7baac8aea331127);
- Bump `@typescript-eslint/eslint-plugin` from `6.10.0` to `6.11.0` (https://github.com/pcaversaccio/createx/pull/31/commits/2390acb96bbf743b5a8e4a04640257c32883bda1);
- Bump `@typescript-eslint/parser` from `6.10.0` to `6.11.0` (https://github.com/pcaversaccio/createx/pull/31/commits/2390acb96bbf743b5a8e4a04640257c32883bda1);

#### 🐶 Cute Animal Picture

image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -67,3 +67,7 @@ forge-artifacts
 # Compiler files
 cache
 out
+
+# Presigning and deployment directories
+presigned-createx-deployment-transaction
+broadcasted-createx-deployment-transaction
```
Comment (mds1) [inline .gitignore:73]: Why are these gitignored, don't we want anyone to be able to deploy the presigned transaction?

**Review by mds1 — COMMENTED** [review]
In general this all seems good. One question I have is: once the code is frozen, how can we ensure we always deploy identical bytecode? For example, perhaps the scripts here should never compile the contract, and instead should read from a build artifact that we commit to the repo

File: .gitignore (hunk 1 above)
Comment (pcaversaccio) [reply .gitignore:73]: That's just temporary until we froze the code and generated the final transaction. I don't want to pollute the commit history with presigned transactions that are actually not valid. Also, it makes it safe for local testing without accidentally pushing a presigned transaction.

**Comment by pcaversaccio** [comment]
> In general this all seems good. One question I have is: once the code is frozen, how can we ensure we always deploy identical bytecode? For example, perhaps the scripts here should never compile the contract, and instead should read from a build artifact that we commit to the repo

Right, so to emphasise, these scripts will be amended/refactored once we freeze the code. The reason why for example the `presign.ts` script compiles is that I haven't added (yet) the `--no-compile` option to the command `npx hardhat run --no-compile scripts/presign.ts`, which makes local testing much easier. I actually plan to add to _all_ `npx hardhat run` commands the `--no-compile` option, for the `presign.ts` script I will hardcode the initcode in a separate file (after confirming with you the validity) and will add an additional check in the script as well as in CI that the `keccak256` hash of the initcode always matches. For the `deploy.ts` script, I will also add the `--no-compile` option to all scripts, will add a sanity check that the `keccak256` hash of the presigned transaction always matches for deploying a raw transaction, and for the `deployNormal` function I will use instead the `loadArtifact(...)` method for deploying and will upload the artifact in a specific folder I can specify in the `hardhat.config.ts` folder:

```ts
const artifact = await deployer.loadArtifact("Greeter");
const contract = await deployer.deploy(artifact);
```

For the verification part, I will also need to upload the build information so it can be used without compilation.

**Comment by mds1** [comment]
> Right, so to emphasise, these scripts will be amended/refactored once we freeze the code. 

Ah this is the key part I was missing. Ok, that all sounds great and this PR LGTM—thanks!

**Review by mds1 — APPROVED** [review]

**Pull request merged**